### ISSUE TITLE
Error 400 response "Mismatching redirect uri"

### ISSUE THREAD

**Issue by timbl** [inrupt/solid-client-authn-js#2891]
### Impacted package

Which packages do you think might be impacted by the bug ?

- [ ] solid-client-authn-browser
- [ ] solid-client-authn-node
- [ ] solid-client-authn-core
- [ ] oidc-client-ext
- [ ] Other (please specify): ...

### To Reproduce

1.
2.
3.
4.

### Environment

Please run

npx envinfo --system --npmPackages --binaries --npmGlobalPackages --browsers

in your project folder and paste the output here:

```
$ npx envinfo --system --npmPackages --binaries --npmGlobalPackages --browsers
```

**Comment by NSeydoux** [comment]
Hi @timbl , sorry for the delayed response! We'll look into this soon.

**Comment by NSeydoux** [comment]
I'm looking into this now, and I started a test to reproduce locally, but since it requires waiting it's going to be a while. I'd be interested in a bit more details form people experiencing this. 

First, what Identity Providers is this happening against?

Then, it would be helpful to have a bit more info about the network log leading to this issue. In particular, I'd be interested in the POST request to the `/register` endpoint, to have the dynamically registered client id and redirect URL, as well as the redirections to the `/authorize` endpoint that causes the issue, to be able to compare the client id and redirect URL.

**Comment by jeff-zucker** [comment]
These are the request headers : 
```
GET /authorize?client_id=a25fc2bd57cb9c3e6f668b7cfc071d0d&redirect_uri=https%3A%2F%2Fjeff-zucker.solidcommunity.net%2F%3Fstate%3D77b54cff18f345188f2aa202e06bfe11&response_type=code&scope=openid%20offline_access%20webid&state=039ee31869214bf2be40fa53921c7c09&code_challenge=XpT9tSAtELBamOSXyLmBxTBDFF1wtzDVzocc-c4trlU&code_challenge_method=S256&prompt=none&response_mode=query HTTP/1.1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cache-Control: no-cache
Connection: keep-alive
DNT: 1
Host: solidcommunity.net
Pragma: no-cache
Referer: https://jeff-zucker.solidcommunity.net/
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: same-site
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36
sec-ch-ua: "Chromium";v="116", "Not)A;Brand";v="24", "Google Chrome";v="116"
sec-ch-ua-arch: "x86"
sec-ch-ua-bitness: "64"
sec-ch-ua-full-version: "116.0.5845.96"
sec-ch-ua-full-version-list: "Chromium";v="116.0.5845.96", "Not)A;Brand";v="24.0.0.0", "Google Chrome";v="116.0.5845.96"
sec-ch-ua-mobile: ?0
sec-ch-ua-model: ""
sec-ch-ua-platform: "Linux"
sec-ch-ua-platform-version: "5.4.0"
sec-ch-ua-wow64: ?0
```

**Comment by NSeydoux** [comment]
> I am getting it against solidcommunity.net. Today I tried it with both chrome and firefox and simply trying to access my pod gives me the 400 error on both. 

Am I correct to say this means you:
- Open https://jeff-zucker.solidcommunity.net/ in your browser
- Click login
- Get the 400 error
?

> These are the request headers :

If you clear local storage, and log in, do you see a request to the `/register` endpoint prior to the request to `/authorize`? If so, the response should include both a client id, client secret and redirect URL (among other things). Do the client ID and redirect URL match those in the request to `authorize`? 

In particular, I think the issue comes from the `state` component in the redirect URL, which should not include dynamic parts like this.

**Comment by jeff-zucker** [comment]
No, I do not get an opportunity to login.  I simply type the URL in the browser, hit enter and get the error.  As I said, clearing the local cache does nothing.  I can not login, logout, or view a page.  I'm not sure how I could see the request to /register, if the browser shows it, I don't know where to find it.

**Comment by jeff-zucker** [comment]
> If you clear local storage, and log in

I am not able to login, the 400 error appears as soon as I enter a URL on my pod in the browser.

**Comment by NSeydoux** [comment]
After a screensharing session yesterday (thanks to @jeff-zucker for making himself available), we determined that the issue appears to come from the way SolidOS sometimes registers the redirect URL at some point during dynamic client registration. One way to fix the issue is to:
- Temporarily disable JS on the browser (to prevent from being redirected)
- Visit your Pod with SolidOS (nothing should happen, because JS is disabled)
- Clear the local storage items prefixed with `solidClientAuthenticationUser:`
- Re-enable JS

This should get users facing the issue sorted out. In the meantime, I'll implement a fix in the library to prevent the bug from happening in SolidOS, but that may lead to issues popping up in SolidOS, so the team will need to be careful when doing the migration.

**Comment by bourgeoa** [comment]
Jeff is there a way to reproduce the error after clearing JavaScript.
I'm wondering if the issue is related to SolidOS or solidcommunity.net
implementation. Do we have the same issue with inrupt.net ?

**Comment by jeff-zucker** [comment]
I have not been able to reproduce the error and have no idea what caused it to begin with.  When I get the error there is no way to login on my own pod on any page.  If I go to your pod and login I can then navigate to my page but when I logout and try to re-login on one of my own pages, same error as before.

One thing I did notice is that when we finally stopped javascript, cleared the localStorage, restarted javascript, went to pod home, logged in ... it redirected me to https://jeff-zucker.solidcommunity.net/#me even though I had not entered the #me part in the location bar.

**Comment by bourgeoa** [comment]
This is a very interesting this. There should never be this extension


Le mer. 23 août 2023, 16:52, Jeff Zucker ***@***.***> a
écrit :

**Comment by NSeydoux** [comment]
I can reproduce the issue by polluting local storage with an invalid redirect URL, and the issue is fixed by the changes introduced in #3103 , so when this change is merged and released you'll be able to upgrade SolidOS and that should get users out of trouble.

**Comment by timbl** [comment]
Sounds good. 
Will you put the condition into the test suite?

**Comment by timbl** [comment]
And will you add the UrIs which mismatched to the error message or console log or both?

**Comment by NSeydoux** [comment]
There is a non-regression test that covers the additional check on the redirect URL, and the offending URL is part of the error message, along with the validation criteria: https://github.com/inrupt/solid-client-authn-js/blob/ac6988c96ebfeaf67a76d59d8c15672c55407cab/packages/browser/src/ClientAuthentication.ts#L68

**Comment by NSeydoux** [comment]
As a continuation of the discussion in https://forum.solidproject.org/t/podbrowser-deprecation-announcement/6212/8:
I’ll check if this fix has been applied to PodBrowser. In the meantime, a solution that works to clear the local storage is to disable Javascript before going to `https://podbrowser.inrupt.com`: it prevents the initial redirect to happen, and allows to remain on the desired origin to clear the storage.

In other words, users of the authentication library shouldn’t hopefully run into this issue if they are on the latest version.

Out of curiosity, could you check the values in storage, and in particular the redirect URL for the registered client? In the past, this issue has been caused by the redirect URL being overwritten with a value including OpenID query parameters, due to a race condition on redirect I haven't been able to reproduce locally because of its random nature.

**Comment by NoelDeMartin** [comment]
Ok thanks for the tip @NSeydoux, it is working now :).

I haven't cleared up the storage though, in case it's useful to debug. What value do you want to see exactly?

This is everything I have in local storage:

![image](https://github.com/inrupt/solid-client-authn-js/assets/1517677/313b48b2-46bb-4c6c-bc27-dc2bce6b4f8c)

This is the value for `issuerConfig:https://solidcommunity.net`:

![image](https://github.com/inrupt/solid-client-authn-js/assets/1517677/8ff5984d-b6ca-49d4-b58b-7fdae5147825)

And this is the value for the second `solidClientAuthenticationUser` (the first one only has a `sessionId`):

![Screenshot from 2023-09-29 13-44-50](https://github.com/inrupt/solid-client-authn-js/assets/1517677/447837e5-fe36-46b7-8f1a-f4b9615d4f71)

**Comment by NSeydoux** [comment]
The `redirectUrl` including `?state=...` is the culprit. There's a race condition somewhere that gets this in storage in replacement of the same redirect URL without the `state` parameter. The 1.17.2 version of `@inrupt/solid-client-authn-*` should prevent this from happening.

**Comment by timbl** [comment]
@NSeydoux For people logging into timbl.com it certainly wasn't a race condition --- anyone whoo left their browser logged in overnight would get the issue and have to 5restart with an incognito mode window.  So very repeatable, you just had to wait. Looking forward to the fix comg through.
